### PR TITLE
exclusive range patterns

### PR DESCRIPTION
**Pull request by oli-obk**
adds `..` patterns to the language under a feature gate (`exclusive_range_pattern`).

This allows turning

``` rust
match i {
    0...9 => {},
    10...19 => {},
    20...29 => {},
    _ => {}
}
```

into 

``` rust
match i {
    0..10 => {},
    10..20 => {},
    20..30 => {},
    _ => {}
}
```

### CODE REVIEW

File: src/librustc/hir/mod.rs [hunk 1]
```diff
@@ -585,6 +586,8 @@ pub enum PatKind {
     Lit(P<Expr>),
     /// A range pattern, e.g. `1...2`
     Range(P<Expr>, P<Expr>),
+    /// A range pattern excluding the end, e.g. `1..2`
+    ExclusiveRange(P<Expr>, P<Expr>),
```
Comment (petrochenkov) [inline src/librustc/hir/mod.rs]: In most cases inclusive and exclusive ranges behave identically, it's probably more convenient to merge them into one variant `Range(P<Expr>, P<Expr>, RangeKind)`.

Comment (oli-obk) [inline src/librustc/hir/mod.rs]: I changed this, it's much less and more readable code now :+1:

**Comment by oli-obk** [comment]
rebased

**Comment by oli-obk** [comment]
I'll bump the error code once this is deemed ready.

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -3652,14 +3653,19 @@ impl<'a> Parser<'a> {
                         let mac = spanned(lo, self.prev_span.hi, Mac_ { path: path, tts: tts });
                         pat = PatKind::Mac(mac);
                     }
-                    token::DotDotDot => {
+                    tok @ token::DotDotDot | tok @ token::DotDot => {
```
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: You also need to add `token::DotDot` to this condition above:
```
                       self.look_ahead(1, |t| match *t {
                           token::OpenDelim(token::Paren) | token::OpenDelim(token::Brace) |
                           token::DotDotDot | token::ModSep | token::Not => false,
```
=>
```
                       self.look_ahead(1, |t| match *t {
                           token::OpenDelim(token::Paren) | token::OpenDelim(token::Brace) |
                           token::DotDotDot | token::DotDot | token::ModSep | token::Not => false,
```

**Comment by brson** [comment]
This PR seems to have gone under the radar for a long time. Can somebody review?

Is this an implementation of an RFC. Is there a tracking issue?

**Comment by brson** [comment]
cc @rust-lang/compiler

**Comment by oli-obk** [comment]
No rfc or tracking issue, I thought it was just an oversight. But I can write an RFC...

**Comment by nikomatsakis** [comment]
There was an RFC, iirc, but @aturon has been having second thoughts on the syntax here, and wanting something more general.

**Comment by nikomatsakis** [comment]
My mistake: I thought this was adding `...` expressions, not `..` patterns.

**Comment by aturon** [comment]
Note: this was not covered by https://github.com/rust-lang/rfcs/pull/1192

**Comment by eddyb** [comment]
Any thoughts on `a..` or `..b`?

**Comment by nikomatsakis** [comment]
Discussed in the @rust-lang/lang meeting:

- This needs a tracking issue and feature-gate.
- We feel OK landing this for now, as long as it is gated, because it seems that in any possible world this is desirable, **but:**
    - we would be unlikely to stabilize until the question of how to handle `...` is resolved
        - my preference is `..=` (which generalizes to `<..<`, `<..=` and so forth)

**Comment by oli-obk** [comment]
added tracking issue, the feature gate was already there.

**Comment by bluss** [comment]
I'll put in my pedantic note that these are "exclusive-end" ranges.

**Comment by oli-obk** [comment]
Rebased. This is ready for r+

**Comment by aturon** [comment]
cc @rust-lang/compiler this PR is ready to go, but needs an active reviewer.

File: src/libsyntax/parse/parser.rs [hunk 3]
```diff
@@ -3527,6 +3526,22 @@ impl<'a> Parser<'a> {
...
 
+    // helper function to decide whether to parse as ident binding or to try to do
+    // something more complex like range patterns
+    fn parse_as_ident(&mut self) -> bool {
+        self.look_ahead(1, |t| match *t {
+            token::OpenDelim(token::Paren) | token::OpenDelim(token::Brace) |
+            token::DotDotDot | token::ModSep | token::Not => Some(false),
+            // ensure slice patterns [a, b.., c] and [a, b, c..] don't go into the
+            // range pattern branch
+            token::DotDot => None,
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: Why is `DotDot` different from `DotDotDot` here?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

**Comment by nikomatsakis** [comment]
Oh, I forgot to ask: can you add some tests targeting the exhaustiveness algorithm? Does it require any changes here?

**Comment by nikomatsakis** [comment]
cc @Manishearth -- are we still grouping changes to the AST together?

**Comment by Manishearth** [comment]
Not at the moment, experimenting with it. ccign some of the folks and landing should be enough

cc @dtolnay @erickt 


feel free to land whenever

**Comment by dtolnay** [comment]
:+1: I always appreciate a cc but no need to batch things like this anymore. Long live macros 1.1!

File: src/libsyntax/parse/parser.rs (hunk 3 above)
Comment (petrochenkov) [inline src/libsyntax/parse/parser.rs]: [Syntactic conflict](https://github.com/rust-lang/rust/issues/23121#issuecomment-154138818) between subslice patterns and exclusive range patterns is finally revealing itself.

**Comment by nikomatsakis** [comment]
@oli-obk at minimum, needs a rebase, but I'd also like to see some tests targeting the exhaustiveness algorithm (I don't think there were any, right?)

**Comment by oli-obk** [comment]
I'll write the tests, it's sensible to have many tests around the various cases here. I just was on vacation and thus didn't get around to answering

**Comment by nikomatsakis** [comment]
@oli-obk no worries. thanks!

**Comment by oli-obk** [comment]
I'm confused... We don't have exhaustiveness checks for integers?

The following doesn't work on the playground:

```rust
    match 42_u8 {
        0...255 => {}
    }
```

I'm not sure how to test exhaustiveness for integers.

**Comment by nikomatsakis** [comment]
@oli-obk yeah I was thinking that later, that maybe this doesn't intersect exhaustiveness because we're not smart enough there yet. I guess this is only usable for integers?

**Comment by oli-obk** [comment]
It also works for floats, there's a test for them, too. Especially for floats this is an important change, as you could not write the same code with an inclusive match

**Comment by oli-obk** [comment]
Travis has passed

**Review by nikomatsakis — APPROVED** [review]

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
One more thing: can we have some tests regarding the syntactic conflict between exclusive range patterns and slice patterns?

@petrochenkov maybe you can supply a few pertinent examples and what you expect as the outcome? :)

File: src/libsyntax/parse/parser.rs (hunk 3 above)
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: It *does* feel like we should change the slice pattern syntax here. However, could we address with simply having a bit more lookahead? In particular, we don't support patterns like `0..`, do we? (Though...I think we probably should... and I don't know what to use for slices instead just now >:/)

File: src/test/compile-fail/pat-range-bad-dots.rs [hunk 4]
```diff
@@ -8,11 +8,9 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
```
Comment (est31) [inline src/test/compile-fail/pat-range-bad-dots.rs]: Note: with #38914 merged in the meantime, you'll either have to rename this test to "feature-gate-exclusive-range-pattern.rs" or add a comment `// gate-test exclusive_range_pattern`. See also [COMPILER_TESTS.md](https://github.com/rust-lang/rust/blob/master/COMPILER_TESTS.md#summary-of-header-commands). Just saying as changing beforehad will prevent a failed bors run and maybe contribute to getting this merged faster.

**Comment by oli-obk** [comment]
Added tests with exclusive range patterns inside slice patterns. For now this works, but if we ever get half open range patterns we'd get collisions.

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@oli-obk where is the tracking issue here?

**Comment by nikomatsakis** [comment]
Ah, https://github.com/rust-lang/rust/issues/37854 -- didn't come up in search for some reason

**Comment by alexcrichton** [comment]
@bors: retry

* that seems like a bug in travis?

**Comment by alexcrichton** [comment]
@bors: retry

* bug in travis?

**Comment by alexcrichton** [comment]
@bors: retry

* sccache wedged https://github.com/rust-lang/rust/issues/39003

**Comment by alexcrichton** [comment]
@bors: retry

* weird travis bug?

**Comment by alexcrichton** [comment]
@bors: retry

* osx linker segfault https://github.com/rust-lang/rust/issues/38878

**Comment by alexcrichton** [comment]
@bors: retry

* osx linker segfaulted https://github.com/rust-lang/rust/issues/38878

**Comment by LU15W1R7H** [comment]
I just came to this issue, because I wanted to write the code you see below.
```rs
match idx {
  0..LOCAL_BOARD_SIZE => LineType::XAxis(idx as u8),
  LOCAL_BOARD_SIZE..2 * LOCAL_BOARD_SIZE => LineType::YAxis((idx - LOCAL_BOARD_SIZE) as u8),
  2 * LOCAL_BOARD_SIZE => LineType::MainDiagonal,
  2 * LOCAL_BOARD_SIZE + 1 => LineType::AntiDiagonal,
  _ => unreachable!(),
}
```

I'm shocked by how many people are against this feature and are concerned with it's introduction, because of potentially wrong usage. I really don't get this. There are perfectly valid places for using this, as my example shows.

I think this is very much related to computer science being 0-indexed instead of 1-indexed, which I consider one of the best decisions. [^1]

In a 0-indexed world, the number of things isn't equal to the last index, it's one more. But it's still useful to think about the number of things. This is why for loops are written as `for i in 0..10`, because we are 0-indexed but we want 10 iterations. So this is useful! This is why we allow for open ranges in for loops. The same applies to matching, this is no different! If we want to match 10 (number of things) elements, this means we want the elements with the indices 0 to (and including 9), so `0..=9`, but this is cumbersome if you are just thinking about the number of things, so `0..10` is a lot better!
In order to match a certain number of things (instead of the indices) we must allow for exclusive end ranges in matches!

[^1]: I actually believe that humanity should have opted for 0-indexing everywhere, also in mathematics, maybe even day-to-day life? But unfortunately history didn't play out this way.

**Pull request merged**